climada/engine/impact.py

```diff
@@ -940,6 +946,172 @@ def _cen_return_imp(imp, freq, imp_th, return_periods):
+        dates : tuple(), optional
+            Tuple of dates to select the events in between. The default is None.
+        coord_exp : (np.ndarray), optional]
+            Selection of exposures coordinates [lat, lon] (in degrees)
+            The default is None.
+
+        Raises
+        ------
+        ValueError
+            If the impact matrix is missing, the eai_exp and aai_agg cannot
+            be updated for a selection of events and/or exposures.
+
+        Returns
+        -------
+        imp : Impact()
+            A new impact object with a selection of events and/or exposures
+
+        """
+
+        if self.imp_mat.shape == (0, 0):
+            raise ValueError("The impact matrix is missing. eai_exp and\
+aai_agg cannot be computed. Recomputed impact.calc with save_mat=True")
+
+        imp = copy.deepcopy(self)
+        nb_events = imp.event_id.size
+        nb_exp = len(imp.coord_exp)
+        
+        sel_exp = np.arange(nb_exp)
+        # filter exposures
+        if isinstance(coord_exp, np.ndarray):
+            sel_exp = [j
+                       for j, coord in enumerate(imp.coord_exp)
+                       if coord in coord_exp
+                      ]
+
+        # filter events by date
+        if dates is None:
+            sel_dt = np.zeros(nb_events, dtype=bool)
+        else:
+            sel_dt = np.ones(nb_events, dtype=bool)
+            date_ini, date_end = dates
+            if isinstance(date_ini, str):
+                date_ini = util_dt.str_to_date(date_ini)
+                date_end = util_dt.str_to_date(date_end)
+            sel_dt &= (date_ini <= imp.date) & (imp.date <= date_end)
+            if not np.any(sel_dt):
+                LOGGER.info('No impact event in date range %s.', dates)
+                return None     
+        sel_dt = list(np.argwhere(sel_dt).reshape(-1))
+        
+        # filter events by id
+        if isinstance(event_ids, list):
+            sel_id = [list(imp.event_id).index(_id) for _id in event_ids]
+        else:
+            sel_id = []
+
+
+        # filter events by name
+        if isinstance(event_names, list):
+            sel_na= [list(imp.event_name).index(name) for name in event_names]
+        else:
+            sel_na = []
+
+        
+        #select events with machting id, name or date field.
+        sel_ev = [idx for idx in set(sel_dt + sel_id+  sel_na)]
+        
+        #if no date, name or id is set, take all events:
+        if (dates, event_ids, event_names) == (None, None, None):
+            sel_ev = np.arange(nb_events)
+
+        # set all attributes that are 'per event', i.e. have a row or a column
+        # of length equal to the number of events (=nb_events), using the 
+        # selection sel_ev.
+        per_event_attr = self.get_per_event_attr(imp)
+        for (attr, value) in imp.__dict__.items():
+            if attr in per_event_attr:
+                if isinstance(value, np.ndarray) and value.ndim == 1 \
```

Comment:
Or, perhaps a wee bit more directly:
```suggestion
        for attr in self.get_per_event_attr(imp):
            value = imp.__getattribute__(attr)
            if isinstance(value, np.ndarray) and value.ndim == 1 \
            ...